File: 2015/1/julia/soln.jl
using Printf

puzzle_input = open(joinpath(dirname(Base.PROGRAM_FILE), "../input")) do file
    read(file, String)
end


function part1(data)
    return count("(", data) - count(")", data)
end


function part2(data)
    position = 0
    target = -1
    current = 0
    sequence = [x == "(" ? 1 : -1  for x in split(data, "")]
    for instruction in sequence
        current += instruction
        position += 1
        if current == target
            break
        end
    end
    return position
end


@printf "Part 1: %s\n" part1(puzzle_input)

@printf "Part 2: %s\n" part2(puzzle_input)

File: 2015/1/python/soln.py
from pathlib import Path

puzzle_input = Path("./input").read_text().strip()


def part1(data):
    return data.count("(") - data.count(")")


def part2(instructions):
    position = 0
    target = -1
    current = 0
    instructions = [1 if c == "(" else -1 for c in instructions]
    for instruction in instructions:
        current += instruction
        position += 1
        if current == target:
            break
    return position


result = part1(puzzle_input)
print(f"Part 1 ends on floor {result}")
result = part2(puzzle_input)
print(f"Part 2 descends into the basement at position {result}")
